ext/neural_snow/ModelTools.jl

```diff
@@ -2,130 +2,481 @@ module ModelTools
+    #not throw a reference error when loaded:
+    let nft = nfeatures,
+        n_z = n,
+        FT = ftype,
+        p_idx = precipitation_index,
+        z_idx = depth_index
+
+        in_scales =
+            (isnothing(in_scale)) ? Vector{FT}(ones(nfeatures)) :
+            Vector{FT}(1 ./ in_scale)
+
+        upperb(pred, input) = paper_upper_bound(p_idx, pred, input)
+        lowerb(pred, input) = paper_lower_bound(z_idx, pred, input)
+        connectfunc(pred, input) =
+            connection_optimized(upperb, lowerb, pred, input)
+
+        # These matrices set up the enforcing structure
+        # for a nonpositive lower bound -z/Δt and a nonnegative
+        # upper bound 1_{P>0}. The 1/Δt factor in the
+        # lower bound will carried in the `scale_and_relus`
+        # matrix and not the boundary function, and the -1 factor
+        # is already built into the below matrix structure
+        # to enable easier setting of the timescale and permit
+        # a reduced structure. See the paper on how to set up
+        # the enforcing structure for alternative bounds.
+        scale_and_relus = Matrix{FT}([1 0 0; 0 1 0; 1 0 -1])
+        get_min = Matrix{FT}([1 1 -1; 0 1 0])
+        final_mul = Matrix{FT}([1 -1])
+
+        #The actual neural-network definition:
+        model = Chain(
+            pred = SkipConnection(
+                Chain(
+                    scale = x -> x .* in_scales,
+                    l1 = Dense(nft, n_z * nft, relu),
+                    l2 = Dense(n_z * nft, nft, elu),
+                    l3 = Dense(nft, 1),
+                ),
+                connectfunc,
+            ),
+            apply_relus = Dense(scale_and_relus, false, relu),
+            apply_upper = Dense(get_min, false, relu),
+            sumlayer = MulLayer(final_mul),
+        )
+        return convert_model(model, FT)
+    end
 end
 
+
 """
-    settimescale!(model, dt; dtype)
+    setoutscale!(model, scale)
+
+Set the output scaling parameter for model usage (i.e. rectifying scaling done on model input).
+This may have unintended results or not work for models not created with `make_model()` or `make_model_paper()`.
 
-Set the timescale parameter for model usage.
+Note: This should only be used for models made with `make_model()` or `make_model_paper()`.
+This scaling constant is combined with a -1 factor that enforces the boundary for speed and memory purposes -
+for this reason, we do not recommend first `extracting`/storing the weight this function refers to,
+and then using that value with this function to `reset` the weight the intial/stored value, as this will have unintended
+consequences without accounting for the extra factor. Only use this function to scale the output 
+of the predictive component by the constant `scale`.
+
+# Arguments
+- `model::Chain`: the neural model to be scaled.
+- `scale::Real`: the scaling parameter to return data to applicable units.
+"""
+function setoutscale!(model::Chain, scale::Real)
```

Comment:
```suggestion
function setoutscale!(model::Chain, scale)
```
You don't need to annotate every type. 

Use type annotations if needed for dispatch or if a reader would find type would be very hard to guess otherwise.